### PR TITLE
Adds support for Numeric types with arbitrary precision

### PR DESCRIPTION
**Pull request by erezsh**
Tested on all supported DBs.

We still need more thorough tests, to cover all db types, and more edge-case values.

### CODE REVIEW

**Comment by sirupsen** [comment]
Would you be able to supply a description of what you're doing here to assist with my review tomorrow? You can either do this by updating the code with comments, or leaving a comment here

I'm finding it fairly hard to follow (and thus review), since we haven't talked as much about this as we did for datetime

**Comment by erezsh** [comment]
Added some comments, hope it helps. If not, find me in the chat and I'll explain it live.

File: tests/test_database_types.py [hunk 1]
```diff
@@ -37,15 +44,15 @@
         ],
         # https://www.postgresql.org/docs/current/datatype-datetime.html
         "datetime_no_timezone": [
-            "timestamp(6) without time zone",
-            "timestamp(3) without time zone",
-            "timestamp(0) without time zone",
+            # "timestamp(6) without time zone",
+            # "timestamp(3) without time zone",
+            # "timestamp(0) without time zone",
```
Comment (sirupsen) [inline tests/test_database_types.py:49]: Before you merge, make sure to uncomment these again 👍🏻

File: tests/test_database_types.py [hunk 2]
```diff
@@ -184,25 +212,37 @@ def _insert_to_table(conn, table, values):
...
     conn.query(insertion_query, None)
     if not isinstance(conn, db.BigQuery):
         conn.query("COMMIT", None)
 
+
 def _drop_table_if_exists(conn, table):
     with suppress(db.QueryError):
         if isinstance(conn, db.Oracle):
             conn.query(f"DROP TABLE {table}", None)
+            conn.query(f"DROP TABLE {table}", None)
```
Comment (sirupsen) [inline tests/test_database_types.py:241]: Why'd you need this twice?

File: data_diff/diff_tables.py [hunk 3]
```diff
@@ -364,11 +364,23 @@ def _validate_and_adjust_columns(self, table1, table2):
...
                 table2._schema[c] = col2.replace(precision=lowest.precision, rounds=lowest.rounds)
 
+            elif isinstance(col1, NumericType):
+                if not isinstance(col2, NumericType):
+                    raise TypeError(f"Incompatible types for column {c}:  {col1} <-> {col2}")
+
+                lowest = min(col1, col2, key=attrgetter("precision"))
+
+                if col1.precision != col2.precision:
+                    logger.warning(f"Using reduced precision {lowest} for column '{c}'. Types={col1}, {col2}")
```
Comment (sirupsen) [inline data_diff/diff_tables.py:379]: For the types, it might be nice to print the database type as well with the type. Not critical right now, but I think we need to get in that habit... Seen so many errors where I'm wondering whether it's from the left or right table 😅

File: data_diff/diff_tables.py [hunk 4]
```diff
@@ -364,11 +364,23 @@ def _validate_and_adjust_columns(self, table1, table2):
...
+
+                table1._schema[c] = col1.replace(precision=lowest.precision)
+                table2._schema[c] = col2.replace(precision=lowest.precision)
```
Comment (sirupsen) [inline data_diff/diff_tables.py:382]: Can you explain why rounding isn't a problem for lower-precision floats and numerics, but is for DateTime? 👂🏻 

E.g. why do you have the problem where if you insert a datetime with `0.88888`, you run into a problem, but not with a `numeric(10, 1)`?

File: data_diff/database.py [hunk 5]
```diff
@@ -212,23 +247,48 @@ def query(self, sql_ast: SqlOrStr, res_type: type):
...
+                return cls(precision=0)
+
+            elif issubclass(cls, Decimal):
+                return cls(precision=numeric_scale)
+
+            assert issubclass(cls, Float)
+            # assert numeric_scale is None
+            return cls(
+                precision=self._convert_db_precision_to_digits(
+                    numeric_precision if numeric_precision is not None else DEFAULT_NUMERIC_PRECISION
```
Comment (sirupsen) [inline data_diff/database.py:281]: Why is 4-bytes float the safe default? Because even if it's 8-bytes, it is "at least" 4-bytes? 

Isn't this only a safe default if it's a float4 and `numeric_precision` isn't in the schema? This seems like code that should be in the driver it's relevant for, not the abstract class, no?

Should we maybe introduce `Double` in addition, so they can be typed out in `NUMERIC_TYPES` for each class?

File: tests/test_database_types.py [hunk 6]
```diff
@@ -58,19 +65,31 @@
...
+            # "timestamp",
+            # "datetime(6)",
+        ],
         # https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
         "float": [
-            # "float",
-            # "double",
-            # "numeric",
+            "float",
+            "double",
```
Comment (sirupsen) [inline tests/test_database_types.py:78]: aWhat about a test where we pass the precision bits to `float`?

File: data_diff/database.py [hunk 7]
```diff
@@ -212,23 +247,48 @@ def query(self, sql_ast: SqlOrStr, res_type: type):
...
+                # We fill in the constant, so we need to ignore it for integers.
+                return cls(precision=0)
+
+            elif issubclass(cls, Decimal):
+                return cls(precision=numeric_scale)
+
+            assert issubclass(cls, Float)
+            # assert numeric_scale is None
+            return cls(
+                precision=self._convert_db_precision_to_digits(
```
Comment (sirupsen) [inline data_diff/database.py:280]: For Postgres, `numeric_precision` in `INFORMATION_SCHEMA.columns` is 24, not 23 as I'd normally expect for a 4-byte float. Do you know why?

This causes a potential problem... For Postgres at least, the precision this function gives, and the documentation, aren't congruent for 4-byte, but is for 8-byte:

```python
math.floor(math.log(2**24, 10))
# => 7

>>> math.floor(math.log(2**53, 10))
15
```

CleanShot 2022-06-21 at 09 49 21@2x

https://www.postgresql.org/docs/current/datatype-numeric.html

File: data_diff/database.py [hunk 8]
```diff
@@ -351,6 +421,9 @@ def normalize_value_by_type(self, value: str, coltype: ColType) -> str:
                 timestamp6 = f"to_char({value}::timestamp(6), 'YYYY-mm-dd HH24:MI:SS.US')"
                 return f"RPAD(LEFT({timestamp6}, {TIMESTAMP_PRECISION_POS+coltype.precision}), {TIMESTAMP_PRECISION_POS+6}, '0')"
 
+        elif isinstance(coltype, NumericType):
+            value = f"{value}::decimal(38, {coltype.precision})"
```
Comment (sirupsen) [inline data_diff/database.py:425]: Why can you assume 38 digits in front of the `.`?

File: data_diff/database.py [hunk 9]
```diff
@@ -362,6 +435,10 @@ class Presto(Database):
         "timestamp": Timestamp,
         # "datetime": Datetime,
     }
+    NUMERIC_TYPES = {
+        "integer": Integer,
+        "real": Float,
```
Comment (sirupsen) [inline data_diff/database.py:440]: I think you're missing [double-precision](https://prestodb.io/docs/current/language/types.html#floating-point) here

File: data_diff/database.py [hunk 10]
```diff
@@ -422,8 +502,24 @@ def _parse_type(self, type_repr: str, datetime_precision: int = None, numeric_pr
...
+                    rounds=False,
+                )
+
+        cls = self.NUMERIC_TYPES.get(type_repr)
+        if cls:
+            if issubclass(cls, Integer):
+                assert numeric_precision is not None
+                return cls(0)
+            elif issubclass(cls, Decimal):
+                return cls(6)
```
Comment (sirupsen) [inline data_diff/database.py:515]: Why is this hardcoded for Decimal?

File: data_diff/database.py [hunk 11]
```diff
@@ -433,6 +529,12 @@ class MySQL(ThreadedDatabase):
         "datetime": Datetime,
         "timestamp": Timestamp,
     }
+    NUMERIC_TYPES = {
+        "double": Float,
+        "float": Float,
+        "decimal": Decimal,
+        "int": Decimal,
```
Comment (sirupsen) [inline data_diff/database.py:536]: Why is `int` decimal?

File: data_diff/database.py [hunk 12]
```diff
@@ -532,14 +645,40 @@ def _parse_type(self, type_repr: str, datetime_precision: int = None, numeric_pr
...
-                return cls(precision=datetime_precision if datetime_precision is not None else DEFAULT_PRECISION,
-                    rounds=self.ROUNDS_ON_PREC_LOSS
+                return cls(
+                    precision=datetime_precision if datetime_precision is not None else DEFAULT_DATETIME_PRECISION,
+                    rounds=self.ROUNDS_ON_PREC_LOSS,
                 )
 
+        cls = {
+            "NUMBER": Decimal,
+            "FLOAT": Float,
```
Comment (sirupsen) [inline data_diff/database.py:655]: What about `LONG`?

File: data_diff/database.py [hunk 13]
```diff
@@ -671,6 +825,10 @@ class Snowflake(Database):
         "TIMESTAMP_LTZ": Timestamp,
         "TIMESTAMP_TZ": TimestampTZ,
     }
+    NUMERIC_TYPES = {
+        "NUMBER": Decimal,
+        "FLOAT": Float,
```
Comment (sirupsen) [inline data_diff/database.py:830]: Snowflake has a ton of aliases, but they are all 4-bytes. I suspect that just works, because `numeric_precision` is just always 53 or whatever? 👍🏻 

CleanShot 2022-06-21 at 09 55 35@2x

File: tests/test_database_types.py [hunk 14]
```diff
@@ -24,7 +31,7 @@
         "2022-05-01 15:10:03.003030",
         "2022-06-01 15:10:05.009900",
     ],
-    "float": [0.0, 0.1, 0.10, 10.0, 100.98],
+    "float": [0.0, 0.1, 0.00188, 0.99999, 0.091919, 0.10, 10.0, 100.98],
```
Comment (sirupsen) [inline tests/test_database_types.py:34]: Some additional values that cause some more tests to fail:

```
     "float": [0.0, 0.1, 0.00188, 0.99999, 0.091919, 0.10, 10.0, 100.98,
         0.001201923076923077, 1/3, 1/5, 1/109, 1/109489, 1/1094893892389, 1/10948938923893289],
```

Failing tests:

* test_types_Snowflake_float_to_Postgres_real
* test_types_Redshift_float8_to_Postgres_real
* test_types_Postgres_real_to_Snowflake_float
* test_types_Postgres_real_to_Redshift_float8
* test_types_Postgres_real_to_Postgres_double_precision
* test_types_Postgres_real_to_MySQL_numeric_65_10_
* test_types_Postgres_double_precision_to_Postgres_real
* test_types_MySQL_numeric_65_10__to_Postgres_real

**Review by sirupsen — COMMENTED** [review]
Really excited about this 😍 This and datetime should by far be the hardest, I hope!

File: tests/test_database_types.py (hunk 2 above)
Comment (erezsh) [reply tests/test_database_types.py:241]: I wish I knew. I just know that removing one of them causes some tests to fail.

File: data_diff/diff_tables.py (hunk 4 above)
Comment (erezsh) [reply data_diff/diff_tables.py:382]: It seems like all databases round their floats, instead of truncating them. Which makes sense as they all use the same FP algorithm.

But maybe we just didn't hit the right edge cases yet.

File: data_diff/database.py (hunk 5 above)
Comment (erezsh) [reply data_diff/database.py:281]: Yes.

> not the abstract class

See the refactor PR, it addresses this concern.

> Should we maybe introduce Double in addition

It isn't necessary for most DBs, because they provide numeric_scale.

For those that don't, yes, it can give us more accuracy. It just doesn't seem like high priority.

File: tests/test_database_types.py (hunk 6 above)
Comment (erezsh) [reply tests/test_database_types.py:78]: Maybe? I didn't make sure we're testing every variation.

File: data_diff/database.py (hunk 7 above)
Comment (erezsh) [reply data_diff/database.py:280]: No idea why..  Double still gives 53..

File: tests/test_database_types.py (hunk 2 above)
Comment (sirupsen) [reply tests/test_database_types.py:241]: 😭

File: data_diff/database.py (hunk 5 above)
Comment (sirupsen) [reply data_diff/database.py:281]: 👍🏻 I think it's worth a comment why this is a safe default, but yeah, I agree with you

File: data_diff/database.py (hunk 10 above)
Comment (erezsh) [reply data_diff/database.py:515]: Presto doesn't provide the numeric_scale. 6 is just a safe value.

File: data_diff/database.py (hunk 13 above)
Comment (erezsh) [reply data_diff/database.py:830]: These aliases are usually "normalized" to a single type by the dbs before stored in INFORMATION_SCHEMA.

File: tests/test_database_types.py (hunk 14 above)
Comment (erezsh) [reply tests/test_database_types.py:34]: Nice. I'll look into it!

Comment (erezsh) [reply tests/test_database_types.py]: Working now.

File: data_diff/database.py (hunk 10 above)
Comment (erezsh) [reply data_diff/database.py]: Changed the way it's handled.

**Comment by erezsh** [comment]
@sirupsen I think we're good. Anything else?

**Review by sirupsen — COMMENTED** [review]
Let's goooo! 

Cut 0.1 when this is in master, and seed a changelog file? Can wait until you wake up tomorrow if you like, since the launch won't be until the morning eastern time I think.
